gatherer.go

```diff
@@ -421,4 +452,90 @@ func (g *gatherer) resolveLookupName(e Entity) string {
 	return e.lookupName()
 }
 
+func (g *gatherer) nameByPath(f *descriptor.FileDescriptorProto, path []int32) (string, error) {
+	const (
+		packagePath     = 2 // FileDescriptorProto.Package
+		messageTypePath = 4 // FileDescriptorProto.MessageType
+		enumTypePath    = 5 // FileDescriptorProto.EnumType
+		servicePath     = 6 // FileDescriptorProto.Service
+
+		messageTypeFieldPath      = 2 // DescriptorProto.Field
+		messageTypeNestedTypePath = 3 // DescriptorProto.NestedType
+		messageTypeEnumTypePath   = 4 // DescriptorProto.EnumType
+		messageTypeOneofDeclPath  = 8 // DescriptorProto.OneofDecl
+	)
+
+	// return fast in case it's the package leading comment
+	packageName := f.GetPackage()
+	if path[0] == packagePath {
+		return packageName, nil
+	}
+
+	// as we're refering to concrete entities, entity type should be followed by
+	// an index number thus always leading to even paths.
+	if len(path)%2 != 0 {
+		return "", errors.New("path must have even elements")
+	}
+
+	// tail call
+	var fn func(interface {
+		GetName() string
+	}, []int32, *[]string) error
+	fn = func(parent interface {
```

Comment:
just `fn :=` should be fine, no?

Reply:
No, because it's used inside the functor itself and will not be discoverable.